File: 05_Sliding_Window/Reverse_Array_in_Groups/reverse_array_in_groups.cpp
// GFG Problem - https://www.geeksforgeeks.org/problems/reverse-array-in-groups0255/1

// Window of size k | TC - O(n), SC - O(1)

void reverseInGroups(vector<int>& arr, int k){
    
    // Get the size of the array
    int n = arr.size();  

    for (int i = 0; i < n; i += k) {
        int left = i;

        // to handle case when k is not multiple of n
        int right = min(i + k - 1, n - 1);

        // reverse the sub-array [left, right]
        while (left < right)  {
            swap(arr[left++], arr[right--]);
        }
    }
}

File: 05_Sliding_Window/Reverse_Array_in_Groups/reverse_array_in_groups.java
// GFG Problem - https://www.geeksforgeeks.org/problems/reverse-array-in-groups0255/1

// Window of size k | TC - O(n), SC - O(1)
public class reverse_array_in_groups {
  
   public void reverseInGroups(int[] arr, int k) {
        int n = arr.length;

        for (int i = 0; i < n; i += k) {
            int left = i;
            int right = Math.min(i + k - 1, n - 1); // handle last group

            while (left < right) {
                int temp = arr[left];
                arr[left] = arr[right];
                arr[right] = temp;
                left++;
                right--;
            }
        }
    }
}